Update pipboy.cc with full pagination

cambragol opened the pull request:
Adds full pagination to Status, Data and Automaps main and sub pages to prevent overflow.

fixes #461 and makes #465 #466 redundant

cambragol committed:
Update pipboy.cc

Adds pagination to Automaps sub-location (elevation) pages to prevent overflow over the map display.

Prevents duplicate entries in main Automaps page list (Fallout Sonora)

fixes #461

cambragol committed:
Update pipboy.cc

removed the elimination of duplicated names as there could be a better solution, and it overlaps with other features

cambragol committed:
Update pipboy.cc for full pagination

Full implementation of pagination for the pipboy. Status, Data, and Automaps, and their subpages can now paginate when entries exceed a range which fits in the page.

The 'Video Archives' page was not given pagination, as it seem unlikely it would be needed, and there currently is a limit to number of videos.

cambragol committed:
Update pipboy.cc

Small fix to add correct number of buttons when moving between automap pages.

phobos2077 commented on src/pipboy.cc:
```diff
@@ -359,14 +367,20 @@ int gPipboyWindowOldFont;
 // 0x66450C
 int _actcnt;
 
+int _loccnt;
```
Would be nice to have readable names for any newly introduced symbols.

cambragol replied:
I was trying to follow existing naming conventions.

phobos2077 commented on src/pipboy.cc:
```diff
@@ -836,939 +924,1014 @@ static void pipboyDrawText(const char* text, int flags, int color)
-                    pipboyDrawText(text2, PIPBOY_TEXT_ALIGNMENT_RIGHT_COLUMN_CENTER, _colorTable[992]);
+      if ((gPipboyMouseX > 459 && a1 != 1027) || a1 == 1026) {
+         if (gPipboyHolodiskLastPage <= _view_page) {
+            if (a1 != 1026) {
```
Would be nice to deobfuscate/decypher variable names for any code that's being rewritten.

cambragol replied:
I'm not really sure what you mean.

phobos2077 replied:
I mean you keep using `a1` name for the variable. Since the code is being rewritten, might as well name it properly.

cambragol replied:
Ah, okay. Clear now.
I guess I was just trying to be the least disruptive I could. I am not a coder, and this is the first C++ I have ever coded, and the first PRs I have done, so renaming a bunch of stuff seemed presumptive.

mikeklaas replied:
My 2c is that it's better to leave de-obfuscation as a separate change to make it easier to review PRs that actually change semantics vs. ones that are just variable renaming.

phobos2077 commented on src/pipboy.cc:
```diff
@@ -836,939 +924,1014 @@ static void pipboyDrawText(const char* text, int flags, int color)
+   const int maxEntriesPerPage = PIPBOY_STATUS_QUESTLIST_LINES;
+
+   int index = 0;
+   int v13 = 0;
```
See above about deobfuscation. Looking at the diff it's hard to tell which code is new and which was just moved around, so it ended up in the diff. So I would treat any code in the diff as new code with some quality standards.

cambragol commented:
Trying to do some code improvements here, but ran into a bug/issue that bothers me.

Original code does a brief 'highlight' of selected buttons (locations, quests, holodisks etc) before switching to the relevant page.

However, Fallout 2 CE seems to load those pages so fast, or in a way, that prevents the highlight from being seen.

My code ran a little roughshod over this original feature, as I was not seeing it working. But now I realize that it is supposed to work. So I need to restore the code, and then figure out how to get it working.

Here is an example: 

```
      if (gPipboyQuestLocationsCount != 0 && gPipboyWindowQuestsCurrentPageCount >= userInput && gPipboyMouseX < 429) {
         soundPlayFile("ib1p1xx1");
         blitBufferToBuffer(_pipboyFrmImages[PIPBOY_FRM_BACKGROUND].getData() + PIPBOY_WINDOW_WIDTH * PIPBOY_WINDOW_CONTENT_VIEW_Y + PIPBOY_WINDOW_CONTENT_VIEW_X,
            PIPBOY_WINDOW_CONTENT_VIEW_WIDTH,
            PIPBOY_WINDOW_CONTENT_VIEW_HEIGHT,
            PIPBOY_WINDOW_WIDTH,
            gPipboyWindowBuffer + PIPBOY_WINDOW_WIDTH * PIPBOY_WINDOW_CONTENT_VIEW_Y + PIPBOY_WINDOW_CONTENT_VIEW_X,
            PIPBOY_WINDOW_WIDTH);
         pipboyWindowRenderQuestLocationList(userInput);
         pipboyWindowRenderHolodiskList(-1);
         windowRefreshRect(gPipboyWindow, & gPipboyWindowContentRect);
         //inputPauseForTocks(200);
         //delay_ms(350);
         _stat_flag = 1;
```

`pipboyWindowRenderQuestLocationList(userInput);` re-reenders the quest location list, with the selected location (userInput) highlighted. However, no matter what I do, delay_ms, inputPause etc., _stat_flag = 1 renders the quest list page instantly and can't be delayed in any way to allow the brief 'highlight' to be seen.

cambragol committed:
Update pipboy.cc

- improved various variable names
- fixed small bug that allowed clicking through quests to quest locations list
- consolidated all bottom page navigation to handlePipboyPageNavigation()

navigation can be seen/tested by lowering these contstants:

// constants for setting lines per page in pagination functions
const int PIPBOY_STATUS_QUEST_LINES = 19;
const int PIPBOY_STATUS_HOLODISK_LINES = 19;
const int PIPBOY_AUTOMAP_LINES = 19;
const int PIPBOY_AUTOMAP_SUB_LINES = 5;
const int PIPBOY_STATUS_QUESTLIST_LINES = 12;
